crates/bevy_render/src/batching/mod.rs

```diff
@@ -208,21 +499,144 @@ pub fn batch_and_prepare_binned_render_phase<BPI, GBBD>(
+                    input_index,
+                    output_index,
+                });
+
+                batch
+                    .get_or_insert(BinnedRenderPhaseBatch {
+                        representative_entity: entity,
+                        instance_range: output_index..output_index,
+                        dynamic_offset: None,
+                    })
+                    .instance_range
+                    .end = output_index + 1;
+            }
+
+            if let Some(batch) = batch {
+                phase.batch_sets.push(smallvec![batch]);
+            }
+        }
+
+        // Prepare unbatchables.
+        for key in &phase.unbatchable_keys {
+            let unbatchables = phase.unbatchable_values.get_mut(key).unwrap();
+            for &entity in &unbatchables.entities {
+                let Some(input_index) =
+                    GFBD::get_binned_batch_input_index(&system_param_item, entity)
+                else {
+                    continue;
+                };
+                let output_index = data_buffer.add() as u32;
+
+                work_item_buffer.push(PreprocessWorkItem {
+                    input_index,
+                    output_index,
+                });
+
+                unbatchables
+                    .buffer_indices
+                    .add(GpuArrayBufferIndex::<GFBD::BufferData> {
+                        index: output_index,
+                        dynamic_offset: None,
+                        element_type: PhantomData,
+                    });
+            }
+        }
+    }
+}
+
+pub fn write_cpu_built_batched_instance_buffers<GBD>(
     render_device: Res<RenderDevice>,
     render_queue: Res<RenderQueue>,
-    gpu_array_buffer: ResMut<GpuArrayBuffer<F::BufferData>>,
-) {
-    let gpu_array_buffer = gpu_array_buffer.into_inner();
-    gpu_array_buffer.write_buffer(&render_device, &render_queue);
-    gpu_array_buffer.clear();
+    cpu_batched_instance_buffer: Option<ResMut<BatchedCpuBuiltInstanceBuffer<GBD::BufferData>>>,
+) where
+    GBD: GetBatchData,
+{
+    if let Some(mut cpu_batched_instance_buffer) = cpu_batched_instance_buffer {
+        cpu_batched_instance_buffer.write_buffer(&render_device, &render_queue);
+    }
+}
+
+pub fn write_gpu_built_batched_instance_buffers<GFBD>(
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    gpu_batched_instance_buffers: Option<
+        ResMut<BatchedGpuBuiltInstanceBuffers<GFBD::BufferData, GFBD::BufferInputData>>,
+    >,
+) where
+    GFBD: GetFullBatchData,
+{
+    let Some(mut gpu_batched_instance_buffers) = gpu_batched_instance_buffers else {
+        return;
+    };
+
+    gpu_batched_instance_buffers
+        .data_buffer
+        .write_buffer(&render_device);
+    gpu_batched_instance_buffers
+        .current_input_buffer
+        .write_buffer(&render_device, &render_queue);
+    // There's no need to write `previous_input_buffer`, as we wrote
```

Comment:
where does it get written/swapped with the current_input_buffer though?

Reply:
Here: https://github.com/bevyengine/bevy/pull/12773/files#diff-4bf3ed03d4129aad9f5678ba19f9b14ee8e3e61d6f6365e82197b01c74468b10R725-R726